### PR TITLE
Reduce overhead from using RaygunBreadcrumb by lazy allocation of CustomData

### PR DESCRIPTION
**Pull request by snakefoot**
Also skip StackTrace probing, when BreadCrumb is prefilled.

### CODE REVIEW

File: Mindscape.Raygun4Net4/Breadcrumbs/RaygunBreadcrumbs.cs [hunk 1]
```diff
@@ -27,11 +27,14 @@ public IEnumerator<RaygunBreadcrumb> GetEnumerator()
     
     public void Record(RaygunBreadcrumb crumb)
     {
-      if (RaygunSettings.Settings.BreadcrumbsLocationRecordingEnabled)
+      if (!ShouldRecord(crumb))
+        return;
```
Comment (phillip-haydon) [inline Mindscape.Raygun4Net4/Breadcrumbs/RaygunBreadcrumbs.cs:31]: 🙏 please put braces around this :)

File: Mindscape.Raygun4Net4/RaygunClient.cs [hunk 2]
```diff
@@ -439,7 +439,7 @@ public void SendInBackground(Func<RaygunMessage> raygunMessage)
     private void StripAndSend(Exception exception, IList<string> tags, IDictionary userCustomData, RaygunIdentifierMessage userInfo, DateTime? currentTime)
     {
       var requestMessage = BuildRequestMessage();
-      var breadcrumbs = _breadcrumbs.ToList();
+      IList<RaygunBreadcrumb> breadcrumbs = BuildBreadCrumbList();
```
Comment (phillip-haydon) [inline Mindscape.Raygun4Net4/RaygunClient.cs:442]: Curious to know why this is better than calling `.ToList()` ?

**Review by phillip-haydon — COMMENTED** [review]
Thank you for the PR. I've reviewed it, generally happy with the changes, just missing braces on the if statement kicking off my OCD :)

Do have 1 question in regards to not using `.ToList();`

File: Mindscape.Raygun4Net4/RaygunClient.cs (hunk 2 above)
Comment (snakefoot) [reply Mindscape.Raygun4Net4/RaygunClient.cs:442]: `_breadcrumbs.ToList()` will allocate an enumerator and a `List`-object. The `BuildBreadCrumbList()` will also allocate an enumerator, but only allocate a `List`-object when breadcrumbs are found.

File: Mindscape.Raygun4Net4/Breadcrumbs/RaygunBreadcrumbs.cs (hunk 1 above)
Comment (snakefoot) [reply Mindscape.Raygun4Net4/Breadcrumbs/RaygunBreadcrumbs.cs]: Braces has been added.

File: Mindscape.Raygun4Net.Core/Breadcrumbs/RaygunBreadcrumb.cs [hunk 3]
```diff
@@ -8,9 +8,8 @@ public class RaygunBreadcrumb
   {
     public RaygunBreadcrumb()
     {
-      CustomData = new Dictionary<string, object>();
       Level = RaygunBreadcrumbLevel.Info;
-      Type = RaygunBreadcrumbType.Manual.ToString();
+      Type = nameof(RaygunBreadcrumbType.Manual);
```
Comment (phillip-haydon) [inline Mindscape.Raygun4Net.Core/Breadcrumbs/RaygunBreadcrumb.cs:12]: Love this, I always forget to do it.

**Review by phillip-haydon — APPROVED** [review]
Thank you for the PR!

**Pull request merged**